### PR TITLE
Add support for multi-node training

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

To be completed

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ltran5991** [comment]
@regisss, Here are some code improvements to correctly run distributed training on Guadi for a normal or a deepspeed run.

https://github.com/huggingface/optimum-habana/blob/multi_node/optimum/habana/training_args.py#L416-L444

### Current Code

``` python
device = torch.device("hpu")
self._n_gpu = 1

if self.deepspeed:
    # deepspeed inits torch.distributed internally
    from transformers.deepspeed import is_deepspeed_available

    if not is_deepspeed_available():
        raise ImportError("--deepspeed requires deepspeed: `pip install deepspeed`.")
    import deepspeed
    import habana_frameworks.torch.distributed.hccl

    deepspeed.init_distributed(dist_backend="hccl")
    logger.info("DeepSpeed is enabled.")
else:
    from habana_frameworks.torch.distributed.hccl import initialize_distributed_hpu

    world_size, rank, self.local_rank = initialize_distributed_hpu()

    if self.local_rank != -1:
        if world_size > hthpu.device_count():
            raise RuntimeError(
                f"world_size is equal to {world_size} but there are only {hthpu.device_count()} devices."
            )
        if not torch.distributed.is_initialized():
            torch.distributed.init_process_group(
                backend="hccl", rank=self.local_rank, world_size=world_size
            )
            logger.info("Enabled distributed run.")
    else:
        logger.info("Single node run.")
```

### Updated Code

Should update the code to match below. This will correctly initialize the HPUs to run a distributed training for a normal and deepspeed configuration. 

Changes:
* Move the `initialize_distributed_hpu()` outside the conditional statement to initialize the OS environments world_size, rank, and local rank for both scenarios.
* Removes `hthpu.device_count()` check as this is limited to the local node and does not work for scaling across multi-nodes.
* `torch.distributed.init_process_group()` rank refers to global_rank and not local_rank

``` python
device = torch.device("hpu")
self._n_gpu = 1

from habana_frameworks.torch.distributed.hccl import initialize_distributed_hpu
world_size, rank, self.local_rank = initialize_distributed_hpu()

if self.deepspeed:
    # deepspeed inits torch.distributed internally
    from transformers.deepspeed import is_deepspeed_available

    if not is_deepspeed_available():
        raise ImportError("--deepspeed requires deepspeed: `pip install deepspeed`.")
    import deepspeed

    deepspeed.init_distributed(dist_backend="hccl")
    logger.info("DeepSpeed is enabled.")
else:
    if self.local_rank != -1:
        if not torch.distributed.is_initialized():
            torch.distributed.init_process_group(
                backend="hccl", rank=rank, world_size=world_size
            )
            logger.info("Enabled distributed run.")
    else:
        logger.info("Single node run.")
```

**Comment by regisss** [comment]
@ltran5991 Thanks a lot for the suggestions! I will commit this and run a few experiments :+1:

**Comment by ltran5991** [comment]
When checking on performance for multi-nodes with Gradient Accumulation Step (GAS) > 1, there is a miscalculation in the logic.

https://github.com/huggingface/optimum-habana/blob/multi_node/optimum/habana/trainer.py#L606

Currently the `start_time_after_warmup` is set incorrectly. The `args.throughput_warmup_steps` works on micro_batch_step when it should work on mini_batch_step. The logic at the end for performance calculation is focused on mini_batch_step. This is inconsistent and a problem for any GAS > 1. 

The current code below a `step` is a micro_batch_step.

### Current Code
``` python
for step, inputs in enumerate(epoch_iterator):
    if args.throughput_warmup_steps > 0 and args.throughput_warmup_steps == epoch * steps_in_epoch + step:
        start_time_after_warmup = time.time()

.................Skip to End Calculation...................

# Warmup steps are removed from the calculation of speed metrics
num_samples_for_speed_metrics = num_train_samples - args.throughput_warmup_steps * total_train_batch_size
num_steps_for_speed_metrics = self.state.max_steps - args.throughput_warmup_steps
metrics = speed_metrics(
    "train",
    start_time,
    num_samples=num_samples_for_speed_metrics,
    num_steps=num_steps_for_speed_metrics,
    start_time_after_warmup=start_time_after_warmup,
)
```

### Example GAS 1 with Current Logic

The example below process 2 samples per step. A full mini_batch_step takes 1 step. The  `num_samples_for_speed_metrics` is calculated to be 96. This is correct as we skip the first 2 steps which processed 4 samples.

``` python

num_train_samples = 100
world_size = 2
train_batch_size = 1
gradient_accumulation_step = 1

# total_train_batch_size  = 1 * 1 * 2 = 2
total_train_batch_size = train_batch_size * gradient_accumulation_step * world_size

# start_time_after_warmup gets set on step 2 which processed 4 samples
args.throughput_warmup_steps = 2

# num_samples_for_speed_metrics = 100 - (2 * 2) = 96
num_samples_for_speed_metrics = num_train_samples - args.throughput_warmup_steps * total_train_batch_size

# actual_samples_skipped  = 2 * 1 * 2 = 4
# calculated_samples_skipped = 2 * 2 = 4
actual_samples_skipped  = args.throughput_warmup_steps * train_batch_size * world_size 
calculated_samples_skipped = args.throughput_warmup_steps * total_train_batch_size

assert(actual_samples_skipped  == calculated_samples_skipped )
```

### Example GAS 2 with Current Logic

The example below process 2 samples per step. A full mini_batch_step takes 2 step. The  `num_samples_for_speed_metrics` is calculated to be 92. This is incorrect as we skip the first 2 steps which processed 4 samples. Now the `start_time_after_warmup` is marked at the wrong timestamp according to the number of samples processed

``` python

num_train_samples = 100
world_size = 2
train_batch_size = 1
gradient_accumulation_step = 2

# total_train_batch_size  = 1 * 2 * 2 = 4
total_train_batch_size = train_batch_size * gradient_accumulation_step * world_size

# start_time_after_warmup gets set on step 2 which processed 4 samples
args.throughput_warmup_steps = 2

# num_samples_for_speed_metrics = 100 - (2 * 4) = 92
num_samples_for_speed_metrics = num_train_samples - args.throughput_warmup_steps * total_train_batch_size

# actual_samples_skipped  = 2 * 1 * 2 = 4
# calculated_samples_skipped = 2 * 4 = 8
actual_samples_skipped  = args.throughput_warmup_steps * train_batch_size * world_size 
calculated_samples_skipped = args.throughput_warmup_steps * total_train_batch_size

assert(actual_samples_skipped  == calculated_samples_skipped )
```


### Updated Code

Should update the code to match below. This will adjust for GAS > 1 and record timestamp at the correct step.

``` python
for step, inputs in enumerate(epoch_iterator):
    if args.throughput_warmup_steps > 0 and (args.throughput_warmup_steps * args.gradient_accumulation_steps) == epoch * steps_in_epoch + step:
        start_time_after_warmup = time.time()

```

### Example GAS 2 with Updated Logic

Now the `start_time_after_warmup` get sets according to mini_batch steps rather than micro_batch steps. This is in sync with the performance calculation logic at the end of training

``` python

num_train_samples = 100
world_size = 2
train_batch_size = 1
gradient_accumulation_step = 2

# total_train_batch_size  = 1 * 2 * 2 = 4
total_train_batch_size = train_batch_size * gradient_accumulation_step * world_size

# start_time_after_warmup gets set on step 4
args.throughput_warmup_steps = 2

# num_samples_for_speed_metrics = 100 - (2 * 4) = 92
num_samples_for_speed_metrics = num_train_samples - args.throughput_warmup_steps * total_train_batch_size

# actual_samples_skipped  = 2 * 2 * 1 * 2 = 8
# calculated_samples_skipped = 2 * 4 = 8
actual_samples_skipped  = args.throughput_warmup_steps * args.gradient_accumulation_steps * train_batch_size * world_size 
calculated_samples_skipped = args.throughput_warmup_steps * total_train_batch_size

assert(actual_samples_skipped  == calculated_samples_skipped )
```

**Comment by regisss** [comment]
Thanks for catching this @ltran5991! I corrected it in my last commit.

File: examples/multi-node-training/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+python gaudi_spawn.py \
+    --hostfile path_to_my_hostfile --use_deepspeed \
+    path_to_my_script.py --args1 --args2 ... --argsN \
+    --deepspeed path_to_my_deepspeed_config
+```
+where `--argX` is an argument of the script to run.
+
+## Setup
+
+Check out the [documentation](https://huggingface.co/docs/optimum/habana/usage_guides/multi_node_training) to know how to set up your Gaudi instances for multi-node runs on premises or on AWS.
```
Comment (libinta) [inline examples/multi-node-training/README.md:14]: https://huggingface.co/docs/optimum/habana/usage_guides/multi_node_training

404 error

File: examples/multi-node-training/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+A `Dockerfile` is provided [here](https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training/Dockerfile) to easily start a multi-node run on AWS instances.
```
Comment (libinta) [inline examples/multi-node-training/README.md:16]: https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training/Dockerfile 404 error

File: examples/multi-node-training/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+## Environment variables
+
+If you need to set environment variables for all nodes, you specify them in a `.deepspeed_env` file which is located in the same folder as the script you are calling. It is formatted as follows:
+```
+env_variable_1_name=value
+env_variable_2_name=value
+...
+```
+
+You can find an example for AWS instances [here](https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training/.deepspeed_env).
```
Comment (libinta) [inline examples/multi-node-training/README.md:41]: https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training/.deepspeed_env 404 error

File: docs/source/usage_guides/multi_node_training.mdx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+More information [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html#efa-start-instances).
+
+
+## Launching a Multi-node Run
+
+Once your Gaudi instances are ready, you need to:
+
+1. Enable password-less SSH on your instances so that they can communicate with each other. [This explains how to do it](https://docs.habana.ai/en/latest/AWS_User_Guides/AWS_Distributed_Training_Multiple_DL1/AWS_Distributed_Training_Multiple_DL1.html#running-distributed-training-over-multiple-dl1-instances).
+2. On AWS, to train through EFA, `hccl_ofi_wrapper` should be installed. [Here is how to do it](https://docs.habana.ai/en/latest/AWS_User_Guides/AWS_Distributed_Training_Multiple_DL1/AWS_Distributed_Training_Multiple_DL1.html#build-and-store-custom-docker-image-for-training).
+3. On AWS, you need to set the following environment variables (the easiest way is to write a `.deepspeed_env` file as described [here](https://huggingface.co/docs/optimum/habana/usage_guides/multi_node_training#environment-variables)):
```
[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/usage_guides/multi_node_training.mdx:78]: This link will work when the PR is merged. It currently returns an error because it is not yet part of the main branch or the documentation.

File: docs/source/usage_guides/multi_node_training.mdx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  - `HCCL_OVER_OFI=1` and `HCCL_OVER_TCP=0`
+  - `LD_LIBRARY_PATH=path_to_hccl_ofi_wrapper:/opt/amazon/openmpi/lib:/opt/amazon/efa/lib` where `path_to_hccl_ofi_wrapper` is the path to the `hccl_ofi_wrapper` folder which you installed in the previous step.
+  - (optional) `HCCL_SOCKET_IFNAME=my_network_interface`. If not set, the first network interface with a name that does not start with `lo` or `docker` will be used. More information [here](https://docs.habana.ai/en/latest/API_Reference_Guides/HCCL_APIs/Using_HCCL.html?highlight=HCCL_SOCKET_IFNAME#hccl-socket-ifname).
+
+To make this easier, we provide a Dockerfile [here](https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training/Dockerfile). You will just have to copy the public key of the leader node in the `authorized_keys` file of all other nodes.
```
[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/usage_guides/multi_node_training.mdx:83]: Same

File: docs/source/usage_guides/multi_node_training.mdx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+If you need to set environment variables for all nodes, you can specify them in a [`.deepspeed_env`](https://www.deepspeed.ai/getting-started/#multi-node-environment-variables) file which should be located in the local path you are executing from or in your home directory. The format is the following:
+```
+env_variable_1_name=value
+env_variable_2_name=value
+...
+```
+
+You can find an example for AWS instances [here](path_to_deepspeed_env).
+
```
[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/usage_guides/multi_node_training.mdx:126]: Same

File: examples/README.md [hunk 7]
```diff
@@ -49,6 +52,27 @@ python gaudi_spawn.py \
...
+```
+where `--argX` is an argument of the script to run with DeepSpeed and `--hostfile` is [a file specifying the addresses and the number of devices to use for each node](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node) such as:
+```
+ip_1 slots=8
+ip_2 slots=8
+...
+ip_n slots=8
+```
+
+You can find more information about multi-node training in the [documentation](https://huggingface.co/docs/optimum/habana/usage_guides/multi_node_training) and in the [`multi-node-training`](https://github.com/huggingface/optimum-habana/tree/main/examples/multi-node-training) folder where a Dockerfile is provided to easily set up your environment.
```
[the remark this answers is not in the document]
Comment (regisss) [reply examples/README.md:73]: Same

File: examples/multi-node-training/README.md (hunk 1 above)
Comment (regisss) [reply examples/multi-node-training/README.md:14]: Same

File: examples/multi-node-training/README.md (hunk 2 above)
Comment (regisss) [reply examples/multi-node-training/README.md:16]: Same

File: examples/multi-node-training/README.md (hunk 3 above)
Comment (regisss) [reply examples/multi-node-training/README.md:41]: Same

File: docs/source/usage_guides/multi_node_training.mdx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+#### 1. Set up an EFA-enabled security group
+
+To allow all instances to communicate with each other, you need to set up a *security group* as described by AWS in step 1 of [this link](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html#efa-start-security).
+Once this is done, it should look as follows:
+<p align="center">
+    <img alt="Rules of the security group" src="https://huggingface.co/datasets/optimum/documentation-images/resolve/main/habana/multi_node_security_group.png" width="800">
+    <em style="color: grey">Security group for multi-node training on AWS DL1 instances</em>
+</p>
```
[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/usage_guides/multi_node_training.mdx:49]: The alignment issue only appears on Github, but in the doc it will be fine. You can check it here: https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_116/en/usage_guides/multi_node_training

[the remark this answers is not in the document]
Comment (regisss) [reply docs/source/usage_guides/multi_node_training.mdx:49]: Ah yes for the security group, if I set it the same way as this picture, then I cannot connect to my instances via SSH. So I just followed what is written on AWS and it works: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa-start.html#efa-start-security
I think the way you set the security group in your doc works if you launch the run via AWS batch but not if you ssh to the instances as I did.

File: examples/multi-node-training/.deepspeed_env [hunk 9]
```diff
@@ -0,0 +1,4 @@
+HCCL_OVER_TCP=0
```
Comment (libinta) [inline examples/multi-node-training/.deepspeed_env:1]: I don't think we need this. did you try without the above ?

Comment (regisss) [reply examples/multi-node-training/.deepspeed_env:1]: Actually I just copied what I found in the model reference example: https://github.com/HabanaAI/Model-References/tree/1.7.1/PyTorch/nlp/pretraining/deepspeed-bert#bert-15b-with-zero1-and-bf16-mixed-precision-on-32-hpus
I will try without :+1:

**Review by libinta — APPROVED** [review]
